x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosRealm.java

```diff
@@ -0,0 +1,189 @@
+                ActionListener.wrap(userPrincipalNameOutToken -> {
+                    if (userPrincipalNameOutToken.v1() != null) {
+                        buildUser(userPrincipalNameOutToken.v1(), userPrincipalNameOutToken.v2(), listener);
+                    } else {
+                        /**
+                         * This is when security context could not be established may be due to ongoing
+                         * negotiation and requires token to be sent back to peer for continuing
+                         * further. We are terminating the authentication process as this is spengo
+                         * negotiation and no other realm can handle this. We can have only one Kerberos
+                         * realm in the system so terminating with RestStatus Unauthorized (401) and
+                         * with 'WWW-Authenticate' header populated with value with token in the form
+                         * 'Negotiate oYH1MIHyoAMK...'
+                         */
+                        String errorMessage = "failed to authenticate user, gss context negotiation not complete";
+                        ElasticsearchSecurityException ese = unauthorized(errorMessage, null);
+                        ese = unauthorizedWithOutputToken(ese, userPrincipalNameOutToken.v2());
+                        listener.onResponse(AuthenticationResult.terminate(errorMessage, ese));
+                    }
+                }, e -> handleException(e, listener)));
+    }
+
+    private void handleException(Exception e, final ActionListener<AuthenticationResult> listener) {
+        if (e instanceof LoginException) {
+            listener.onResponse(AuthenticationResult.terminate("failed to authenticate user, service login failure",
+                    unauthorized(e.getLocalizedMessage(), e)));
+        } else if (e instanceof GSSException) {
+            listener.onResponse(AuthenticationResult.terminate("failed to authenticate user, gss context negotiation failure",
+                    unauthorized(e.getLocalizedMessage(), e)));
+        } else {
+            listener.onFailure(e);
+        }
+    }
+
+    private void buildUser(final String username, final String outToken, final ActionListener<AuthenticationResult> listener) {
+        // if outToken is present then it needs to be communicated with peer, add it to
+        // response header in thread context.
+        if (Strings.hasText(outToken)) {
+            threadPool.getThreadContext().addResponseHeader(WWW_AUTHENTICATE, NEGOTIATE_AUTH_HEADER_PREFIX + outToken);
+        }
+        final User user = (userPrincipalNameToUserCache != null) ? userPrincipalNameToUserCache.get(username) : null;
+        if (user != null) {
+            listener.onResponse(AuthenticationResult.success(user));
+        } else {
+            /**
+             * TODO: bizybot AD/LDAP user lookup if lookup realm configured
```

Comment:
Quick comment, not a review issue: This TODO is in the wrong spot - you'll want to handle lookup-realms regardless of the cache status.